### PR TITLE
Split rsemVm into a separate report object type

### PR DESCRIPTION
**Pull request by haxscramper**
Long overdue refactoring, no intentional functionality changes, only cleanup for the reporting system implementation. Add the `VMReport` report object, new "VM" category and split mixed semantic analysis report kinds into more manageable chunks.

### CODE REVIEW

File: compiler/ast/report_enums.nim [hunk 1]
```diff
@@ -1035,13 +1056,52 @@ const
...
+      repTraceKinds +
+      repErrorKinds +
+      repDataPassKinds +
+      repFatalKinds +
+      { repNone }
+    )
+
+    assert(
+      len(diff) == 0,
+      "Severity grouping is missins some elements: " & $diff
```
Comment (saem) [inline compiler/ast/report_enums.nim]:
```suggestion
      "Severity grouping is missing some elements: " & $diff
```

**Review by saem — APPROVED** [review]
one typo, otherwise it's good to go.

File: compiler/front/options.nim [hunk 2]
```diff
@@ -665,7 +665,10 @@ func isEnabled*(conf: ConfigRef, report: ReportKind): bool =
           result = true
 
         of repTraceKinds:
-          result = true
+          # Semantic trace kinds are enabled by default and probably should
+          # not be changed - but these reports might (in theory) be
+          # modified at runtime.
+          result = report in conf.notes
```
Comment (zerbina) [inline compiler/front/options.nim:671]: Could you list this (small functional) change in the commit message?

**Review by zerbina — COMMENTED** [review]
Looks good to me. Thanks!

---

There are some pure style changes in `cli_reporter` that are unrelated to this PR (indentation of `case` statements, line breaks for strings, etc.). Would you be okay with moving those to a separate PR? It'd make traversing the history a bit easier.

I don't consider this to be a blocker however - feel free to disregard my request.

File: compiler/front/options.nim (hunk 2 above)
Comment (haxscramper) [reply compiler/front/options.nim:671]: Sorry, I just noticed your review as well (together with a notification about PR merge), but this is not a functional change, just part of the reporting enum sets was tangled together in a weird manner and there were two bugs cancelling each other out ((1)some reports were never added to the allowed `notes` and (2) for trace reports writing was always allowed).

Comment (haxscramper) [reply compiler/front/options.nim:671]: Added comment just documents the logic behind the check should have been added there several months ago

Comment (zerbina) [reply compiler/front/options.nim:671]: Thank you for clarifying.

**Pull request merged**